File: handlers/session_namer.py
"""
Session auto-namer + color — Stop & UserPromptSubmit handler.

On the first Stop event of a session, spawns a background process that:
1. Reads the session transcript to get the first user message
2. Calls Haiku to generate a 2-4 word kebab-case title + prompt bar color
3. Stores in ~/.claude/data/session-titles.json (external registry)

On UserPromptSubmit, if a color has been assigned but not yet applied,
injects a one-time hint so the model can suggest `/color <name>`.

Non-blocking: spawns background process, returns ALLOW immediately.
Fail-open: any error -> silently skip, never block Claude Code.

Avoids known pitfalls:
- 64KB truncation bug (#33165): uses external JSON registry, not session JSONL
- Resume name loss (#26240): external registry persists independently
"""

from __future__ import annotations

import fcntl
import json
import logging
import os
import subprocess
import sys
from datetime import UTC, datetime

from .base import ALLOW, HOME, HookResult, text_result

logger = logging.getLogger(__name__)

REGISTRY_PATH = os.path.join(HOME, ".claude", "data", "session-titles.json")
VALID_COLORS = {"red", "blue", "green", "yellow", "purple", "orange", "pink", "cyan"}


def _load_registry() -> dict:
    try:
        with open(REGISTRY_PATH) as f:
            return json.load(f)
    except Exception:
        return {}


def handle(
    event_type: str,
    tool_name: str,
    tool_input: dict,
    raw_input: str,
) -> HookResult:
    """Name the session on first Stop event (non-blocking)."""
    if os.environ.get("CLAUDE_SESSION_NAMER", "1") == "0":
        return ALLOW

    try:
        # Resolve session_id
        session_id = os.environ.get("CLAUDE_SESSION_ID", "")
        if not session_id:
            try:
                data = json.loads(raw_input) if raw_input else {}
                session_id = data.get("session_id", "")
            except Exception:  # noqa: S110
                pass

        if not session_id:
            return ALLOW

        # Already named? Skip.
        registry = _load_registry()
        if session_id in registry:
            return ALLOW

        # Spawn background namer process
        code = (
            "import os, sys, json, glob, fcntl\n"
            "from datetime import datetime, timezone\n"
            "\n"
            "HOME = os.path.expanduser('~')\n"
            "REGISTRY = os.path.join(HOME, '.claude', 'data', 'session-titles.json')\n"
            f"session_id = {session_id!r}\n"
            "\n"
            "# Find transcript\n"
            "pattern = os.path.join(HOME, '.claude', 'projects', '**', f'{session_id}.jsonl')\n"
            "matches = glob.glob(pattern, recursive=True)\n"
            "if not matches:\n"
            "    sys.exit(0)\n"
            "\n"
            "# Extract first user message\n"
            "first_message = ''\n"
            "try:\n"
            "    with open(matches[0]) as f:\n"
            "        for line in f:\n"
            "            line = line.strip()\n"
            "            if not line:\n"
            "                continue\n"
            "            entry = json.loads(line)\n"
            "            msg_obj = entry.get('message', {}) or {}\n"
            "            role = entry.get('type', '') or msg_obj.get('role', '')\n"
            "            if role == 'user':\n"
            "                msg = msg_obj\n"
            "                content = msg.get('content', '')\n"
            "                if isinstance(content, list):\n"
            "                    for block in content:\n"
            "                        if isinstance(block, dict) and block.get('type') == 'text':\n"
            "                            first_message = block.get('text', '')\n"
            "                            break\n"
            "                elif isinstance(content, str):\n"
            "                    first_message = content\n"
            "                if first_message:\n"
            "                    break\n"
            "except Exception:\n"
            "    sys.exit(0)\n"
            "\n"
            "if not first_message.strip():\n"
            "    sys.exit(0)\n"
            "\n"
            "# Call Haiku via claude CLI (inherits OAuth auth)\n"
            "import subprocess as _sp\n"
            "try:\n"
            "    prompt = ('Generate a session title and pick a prompt-bar color.\\n'\n"
            "              'Title: 2-4 word kebab-case, verb-first, max 30 chars.\\n'\n"
            "              'Color: pick ONE from [red,blue,green,yellow,purple,orange,pink,cyan] '\n"  # noqa: E501
            "              'that matches the task mood/domain.\\n'\n"
            '              \'Return ONLY JSON: {"title":"...","color":"..."}\\n\\n\'\n'
            "              f'User message: {first_message[:500]}')\n"
            "    r = _sp.run(\n"
            "        ['claude', '-p', prompt, '--model', 'haiku', '--output-format', 'text',\n"
            "         '--no-session-persistence'],\n"
            "        capture_output=True, text=True, timeout=120,\n"
            "        env={**os.environ, 'CTX_SUPERVISOR_LEVEL': 'off',\n"
            "             'CLAUDE_SESSION_NAMER': '0'},\n"
            "    )\n"
            "    raw = r.stdout.strip()\n"
            "    # Strip markdown code fences (```json ... ```)\n"
            "    import re as _re\n"
            "    m = _re.search(r'\\{[^}]*\"title\"[^}]*\\}', raw)\n"
            "    if m:\n"
            "        raw = m.group()\n"
            "    try:\n"
            "        parsed = json.loads(raw)\n"
            "        title = parsed.get('title', '').strip()\n"
            "        color = parsed.get('color', '').strip().lower()\n"
            "    except Exception:\n"
            "        title = raw.strip()\n"
            "        color = ''\n"
            "    valid = {'red','blue','green','yellow','purple','orange','pink','cyan'}\n"
            "    if color not in valid:\n"
            "        color = ''\n"
            "except Exception:\n"
            "    sys.exit(0)\n"
            "\n"
            "if not title:\n"
            "    sys.exit(0)\n"
            "\n"
            "# Write to registry with file lock\n"
            "os.makedirs(os.path.dirname(REGISTRY), exist_ok=True)\n"
            "try:\n"
            "    with open(REGISTRY, 'a+') as f:\n"
            "        fcntl.flock(f, fcntl.LOCK_EX)\n"
            "        f.seek(0)\n"
            "        content = f.read().strip()\n"
            "        registry = json.loads(content) if content else {}\n"
            "        created = datetime.now(timezone.utc).isoformat()\n"
            "        registry[session_id] = {\n"
            "            'title': title, 'color': color, 'created_at': created}\n"
            "        f.seek(0)\n"
            "        f.truncate()\n"
            "        json.dump(registry, f, indent=2)\n"
            "        fcntl.flock(f, fcntl.LOCK_UN)\n"
            "except Exception:\n"
            "    pass\n"
        )

        subprocess.Popen(  # noqa: S603
            [sys.executable, "-c", code],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
    except Exception:  # noqa: S110
        pass  # fail-open: never block Claude Code

    return ALLOW


def batch_rename_sessions(session_paths: list[str]) -> dict[str, dict]:
    """Batch-name up to 50 sessions using a single RLM call.

    Extracts the first and last user messages from each session transcript,
    sends them all as context to RLM, and gets back titles + colors.

    Falls back to individual naming on RLM failure.

    Args:
        session_paths: List of absolute paths to session JSONL files (max 50).

    Returns:
        Dict mapping session_id -> {"title": str, "color": str, "status": str}.
    """
    import re
    import sys as _sys

    _sys.path.insert(0, "/Users/joneshong/workshop/core")

    from src.shared.rlm_engine import RLMConfig, RLMEngine

    session_paths = session_paths[:50]
    results: dict[str, dict] = {}

    # Extract first + last user messages from each session
    session_contexts: list[dict] = []
    for path in session_paths:
        session_id = os.path.basename(path).replace(".jsonl", "")
        first_msg = ""
        last_msg = ""
        try:
            with open(path) as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        entry = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    msg_obj = entry.get("message", {}) or {}
                    role = entry.get("type", "") or msg_obj.get("role", "")
                    if role != "user":
                        continue
                    content = msg_obj.get("content", "")
                    if isinstance(content, list):
                        for block in content:
                            if isinstance(block, dict) and block.get("type") == "text":
                                text = block.get("text", "")
                                break
                        else:
                            text = ""
                    elif isinstance(content, str):
                        text = content
                    else:
                        text = ""
                    if text.strip():
                        if not first_msg:
                            first_msg = text[:300]
                        last_msg = text[:300]
        except Exception:  # noqa: S110
            pass  # skip unreadable session files; continue to next path

        if first_msg:
            session_contexts.append(
                {
                    "session_id": session_id,
                    "first_message": first_msg,
                    "last_message": last_msg if last_msg != first_msg else "",
                }
            )

    if not session_contexts:
        return results

    # Build context for RLM
    context_str = json.dumps(session_contexts, ensure_ascii=False, indent=2)

    prompt = (
        f"為以下 {len(session_contexts)} 個 Claude Code session 生成標題和顏色。\n\n"
        "規則:\n"
        "- title: 2-4 word kebab-case, verb-first, max 30 chars\n"
        "- color: 從 [red,blue,green,yellow,purple,orange,pink,cyan] 選一個匹配任務氛圍\n\n"
        "以 JSON 回覆:\n"
        '{"sessions": [{"session_id": "...", "title": "...", "color": "..."}, ...]}\n\n'
        "FINAL() 包住你的 JSON。"
    )

    config = RLMConfig(
        model="haiku",
        sub_model="haiku",
        max_iterations=3,
        max_timeout_secs=45.0,
        max_depth=1,
    )
    engine = RLMEngine(config)

    try:
        rlm_result = engine.completion(prompt=prompt, context=context_str)

        if rlm_result.status != "ok":
            return results

        raw = rlm_result.response
        raw = re.sub(r"```(?:json)?\s*", "", raw)
        raw = re.sub(r"```\s*", "", raw)

        match = re.search(r"\{.*\}", raw, re.DOTALL)
        if not match:
            return results

        data = json.loads(match.group())
        sessions_list = data.get("sessions", [])

        # Write all to registry at once
        os.makedirs(os.path.dirname(REGISTRY_PATH), exist_ok=True)
        try:
            with open(REGISTRY_PATH, "a+") as f:
                fcntl.flock(f, fcntl.LOCK_EX)
                f.seek(0)
                content = f.read().strip()
                registry = json.loads(content) if content else {}

                for entry in sessions_list:
                    sid = entry.get("session_id", "")
                    title = entry.get("title", "").strip()
                    color = entry.get("color", "").strip().lower()
                    if not sid or not title:
                        continue
                    if color not in VALID_COLORS:
                        color = ""
                    registry[sid] = {
                        "title": title,
                        "color": color,
                        "created_at": datetime.now(UTC).isoformat(),
                    }
                    results[sid] = {"title": title, "color": color, "status": "ok"}

                f.seek(0)
                f.truncate()
                json.dump(registry, f, indent=2)
                fcntl.flock(f, fcntl.LOCK_UN)
        except Exception:
            logger.debug("session_namer: failed to write registry", exc_info=True)

    except Exception:
        logger.debug("session_namer: batch naming failed, returning partial results", exc_info=True)

    return results


def handle_color_hint(
    event_type: str,
    tool_name: str,
    tool_input: dict,
    raw_input: str,
) -> HookResult:
    """Inject one-time color hint on UserPromptSubmit."""
    if os.environ.get("CLAUDE_SESSION_NAMER", "1") == "0":
        return ALLOW

    try:
        session_id = os.environ.get("CLAUDE_SESSION_ID", "")
        if not session_id:
            return ALLOW

        # Fast check: already hinted this session?
        marker = f"/tmp/.claude-color-hinted-{session_id}"  # noqa: S108
        if os.path.exists(marker):
            return ALLOW

        registry = _load_registry()
        entry = registry.get(session_id, {})
        color = entry.get("color", "")

        if color and color in VALID_COLORS:
            # Create marker — hint only once per session
            with open(marker, "w") as f:
                f.write(color)
            return text_result(
                f"Session color assigned: {color} — suggest user run `/color {color}`"
            )
    except Exception:  # noqa: S110
        pass

    return ALLOW

File: stations/hook-dispatcher/internal/handlers/session_namer.go
// Package handlers — session_namer.go
// Stop + UserPromptSubmit handler — session auto-namer + color hint.
//
// On the first Stop event of a session, runs a background goroutine (Go in-process) that:
//  1. Reads the session transcript to get the first user message
//  2. Calls Haiku via claude CLI to generate a 2-4 word kebab-case title + color
//  3. Stores in ~/.claude/data/session-titles.json (external registry, file-locked)
//
// On UserPromptSubmit, if a color has been assigned but not yet applied,
// injects a one-time hint so the model can suggest /color <name>.
//
// Non-blocking: spawns goroutine, returns Allow immediately.
// Fail-open: any error -> silently skip, never block Claude Code.
package handlers

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
)

var sessionNamerValidColors = map[string]struct{}{
	"red": {}, "blue": {}, "green": {}, "yellow": {},
	"purple": {}, "orange": {}, "pink": {}, "cyan": {},
}

func init() {
	core.Register("Stop", core.Entry{
		Matcher:    "",
		Handler:    sessionNamerHandle,
		Critical:   false,
		ModuleName: "session_namer",
	})
	core.Register("UserPromptSubmit", core.Entry{
		Matcher:    "",
		Handler:    sessionNamerHandleColorHint,
		Critical:   false,
		ModuleName: "session_namer",
	})
}

// sessionNamerHandle names the session on the first Stop event (non-blocking).
func sessionNamerHandle(_, _ string, _ map[string]any, rawInput string) core.HookResult {
	if os.Getenv("CLAUDE_SESSION_NAMER") == "0" {
		return core.Allow()
	}

	sessionID := os.Getenv("CLAUDE_SESSION_ID")
	if sessionID == "" {
		if strings.TrimSpace(rawInput) != "" {
			var parsed map[string]any
			if err := json.Unmarshal([]byte(rawInput), &parsed); err == nil {
				sessionID, _ = parsed["session_id"].(string)
			}
		}
	}
	if sessionID == "" {
		return core.Allow()
	}

	// Already named? Skip.
	if sessionNamerIsNamed(sessionID) {
		return core.Allow()
	}

	home, _ := os.UserHomeDir()
	// Go in-process — no Python middleman, no fork overhead.
	go sessionNamerRunInProc(sessionID, home)

	return core.Allow()
}

// sessionNamerHandleColorHint injects a one-time color hint on UserPromptSubmit.
func sessionNamerHandleColorHint(_, _ string, _ map[string]any, _ string) core.HookResult {
	if os.Getenv("CLAUDE_SESSION_NAMER") == "0" {
		return core.Allow()
	}

	sessionID := os.Getenv("CLAUDE_SESSION_ID")
	if sessionID == "" {
		return core.Allow()
	}

	marker := fmt.Sprintf("/tmp/.claude-color-hinted-%s", sessionID)
	if _, err := os.Stat(marker); err == nil {
		// Already hinted
		return core.Allow()
	}

	registry := sessionNamerLoadRegistry()
	entry, ok := registry[sessionID]
	if !ok {
		return core.Allow()
	}
	entryMap, ok := entry.(map[string]any)
	if !ok {
		return core.Allow()
	}
	color, _ := entryMap["color"].(string)

	if _, valid := sessionNamerValidColors[color]; !valid || color == "" {
		return core.Allow()
	}

	// Write marker — hint only once per session
	_ = os.WriteFile(marker, []byte(color), 0o644)
	return core.TextResult(fmt.Sprintf("Session color assigned: %s — suggest user run `/color %s`", color, color))
}

func sessionNamerIsNamed(sessionID string) bool {
	registry := sessionNamerLoadRegistry()
	_, exists := registry[sessionID]
	return exists
}

func sessionNamerLoadRegistry() map[string]any {
	home, _ := os.UserHomeDir()
	registryPath := filepath.Join(home, ".claude", "data", "session-titles.json")
	data, err := os.ReadFile(registryPath)
	if err != nil {
		return map[string]any{}
	}
	var registry map[string]any
	if err := json.Unmarshal(data, &registry); err != nil {
		return map[string]any{}
	}
	return registry
}

// sessionNamerRunInProc is the Go in-process replacement for sessionNamerBuildCode.
// Previously this logic was inlined as Python and executed via `python3 -c <code>`.
// Now it runs directly in a goroutine — no Python fork overhead.
//
// Steps (mirrors the Python inline worker exactly):
//  1. glob ~/.claude/projects/**/{sessionID}.jsonl → find transcript
//  2. read transcript → extract first user message text
//  3. build prompt → exec claude CLI (haiku) → parse JSON response
//  4. write title+color to ~/.claude/data/session-titles.json with syscall.Flock
func sessionNamerRunInProc(sessionID, home string) {
	registryPath := filepath.Join(home, ".claude", "data", "session-titles.json")

	// 1. Find transcript via glob
	pattern := filepath.Join(home, ".claude", "projects", "**", sessionID+".jsonl")
	matches, err := filepath.Glob(pattern)
	if err != nil || len(matches) == 0 {
		// filepath.Glob doesn't support ** — use Walk instead
		transcriptPath := sessionNamerFindTranscript(home, sessionID)
		if transcriptPath == "" {
			return
		}
		matches = []string{transcriptPath}
	}

	// 2. Extract first user message from transcript
	firstMessage := sessionNamerExtractFirstUserMessage(matches[0])
	if strings.TrimSpace(firstMessage) == "" {
		return
	}
	if len(firstMessage) > 500 {
		firstMessage = firstMessage[:500]
	}

	// 3. Call Haiku via claude CLI
	prompt := "Generate a session title and pick a prompt-bar color.\n" +
		"Title: 2-4 word kebab-case, verb-first, max 30 chars.\n" +
		"Color: pick ONE from [red,blue,green,yellow,purple,orange,pink,cyan] " +
		"that matches the task mood/domain.\n" +
		"Return ONLY JSON: {\"title\":\"...\",\"color\":\"...\"}\n\n" +
		"User message: " + firstMessage

	env := os.Environ()
	// Suppress recursion + context supervisor overhead
	env = sessionNamerReplaceEnv(env, "CTX_SUPERVISOR_LEVEL", "off")
	env = sessionNamerReplaceEnv(env, "CLAUDE_SESSION_NAMER", "0")

	r := core.RunCmdWithEnv(
		[]string{"claude", "-p", prompt, "--model", "haiku", "--output-format", "text", "--no-session-persistence"},
		"", 120*time.Second, "", env,
	)
	if r == nil || r.ExitCode != 0 {
		return
	}

	raw := strings.TrimSpace(r.Stdout)
	if raw == "" {
		return
	}

	// Extract JSON object from response (may have preamble text)
	jsonPat := regexp.MustCompile(`\{[^}]*"title"[^}]*\}`)
	if m := jsonPat.FindString(raw); m != "" {
		raw = m
	}

	var parsed struct {
		Title string `json:"title"`
		Color string `json:"color"`
	}
	title := ""
	color := ""
	if err := json.Unmarshal([]byte(raw), &parsed); err == nil {
		title = strings.TrimSpace(parsed.Title)
		color = strings.ToLower(strings.TrimSpace(parsed.Color))
	} else {
		title = strings.TrimSpace(raw)
	}
	if _, valid := sessionNamerValidColors[color]; !valid {
		color = ""
	}
	if title == "" {
		return
	}

	// 4. Write to registry with file lock (mirrors fcntl.flock in Python)
	if err := os.MkdirAll(filepath.Dir(registryPath), 0o755); err != nil {
		return
	}
	f, err := os.OpenFile(registryPath, os.O_RDWR|os.O_CREATE, 0o644)
	if err != nil {
		return
	}
	defer f.Close()

	// Exclusive lock — mirrors fcntl.LOCK_EX
	if err := syscall.Flock(int(f.Fd()), syscall.LOCK_EX); err != nil {
		return
	}
	defer syscall.Flock(int(f.Fd()), syscall.LOCK_UN) //nolint:errcheck

	// Read existing registry
	registry := map[string]any{}
	info, _ := f.Stat()
	if info != nil && info.Size() > 0 {
		data, _ := os.ReadFile(registryPath)
		if len(data) > 0 {
			_ = json.Unmarshal(data, &registry)
		}
	}

	// Python datetime.now(UTC).isoformat() emits "+00:00"; match that
	// instead of Go's default "Z" so string comparisons match registry
	// entries written by legacy Python code paths.
	createdAt := time.Now().UTC().Format("2006-01-02T15:04:05-07:00")
	registry[sessionID] = map[string]any{
		"title":      title,
		"color":      color,
		"created_at": createdAt,
	}

	out, err := json.MarshalIndent(registry, "", "  ")
	if err != nil {
		return
	}

	if err := f.Truncate(0); err != nil {
		return
	}
	if _, err := f.Seek(0, 0); err != nil {
		return
	}
	_, _ = f.Write(out)
}

// sessionNamerFindTranscript walks ~/.claude/projects/ to find {sessionID}.jsonl.
// filepath.Glob does not support **, so we use filepath.WalkDir.
func sessionNamerFindTranscript(home, sessionID string) string {
	projectsDir := filepath.Join(home, ".claude", "projects")
	target := sessionID + ".jsonl"
	found := ""
	errStop := fmt.Errorf("stop")
	_ = filepath.WalkDir(projectsDir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.Name() == target {
			found = path
			return errStop // short-circuit
		}
		return nil
	})
	return found
}

// sessionNamerExtractFirstUserMessage reads the JSONL transcript and returns
// the text content of the first user message.
func sessionNamerExtractFirstUserMessage(transcriptPath string) string {
	f, err := os.Open(transcriptPath)
	if err != nil {
		return ""
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var entry map[string]any
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}
		// Determine role: entry.type == "user" or message.role == "user"
		msgObj, _ := entry["message"].(map[string]any)
		role, _ := entry["type"].(string)
		if role == "" && msgObj != nil {
			role, _ = msgObj["role"].(string)
		}
		if role != "user" {
			continue
		}
		if msgObj == nil {
			continue
		}
		content := msgObj["content"]
		switch c := content.(type) {
		case string:
			if strings.TrimSpace(c) != "" {
				return c
			}
		case []any:
			for _, block := range c {
				bMap, ok := block.(map[string]any)
				if !ok {
					continue
				}
				if bMap["type"] == "text" {
					if text, ok := bMap["text"].(string); ok && strings.TrimSpace(text) != "" {
						return text
					}
				}
			}
		}
	}
	return ""
}

// sessionNamerReplaceEnv replaces or appends a key=value in an os.Environ() slice.
func sessionNamerReplaceEnv(env []string, key, value string) []string {
	prefix := key + "="
	for i, e := range env {
		if strings.HasPrefix(e, prefix) {
			env[i] = prefix + value
			return env
		}
	}
	return append(env, prefix+value)
}
